### DISCUSSION TITLE
Can't run PrismaClient inside Next.js middleware

### DISCUSSION THREAD

**Discussion by waylio** [prisma/prisma#12602] [Q&A]
Hi Folks,

How to do a Prisma call inside a Next.js middleware. 

```
    const store = await prisma.store.findUnique({
      include: { theme: true, products: false, },
      where: {
        title: hostname?.split('.')[0]
      }
    })


    url.pathname = `/_sites/${store.theme.title}/${currentHost}${pathname}`;
    return NextResponse.rewrite(url);

```

![image](https://user-images.githubusercontent.com/91939404/161047733-d786cdd2-72b8-4199-99a4-dbe4c3bec0a6.png)

**Comment by austincrim** [comment]
Hey @waylio 👋 ,

Because Next.js middleware are built on Cloudflare Workers, you will need to use the Prisma Data Proxy if you want to use Prisma inside middleware. You can read complete instructions on using the Data Proxy [in our docs](https://www.prisma.io/docs/concepts/data-platform/data-proxy#using-the-data-proxy-in-a-prisma-application).

Otherwise, you could call a regular Next.js API route from your middleware and use the Prisma Client there.

**Reply by narviii** [reply]
Yeah, but the API route goes through middleware. It would be an infinite loop, no?

**Reply by abdessamadely** [reply]
After testing it on my side, it doesn't work. I don't know why, but it seems that calls going out from the middleware are being ignored.

**Reply by gcb** [reply]
Is this still the case? can someone provide an updated doc link? Original answer 404 (lands on doc homepage)

only found the term "proxy" on https://www.prisma.io/docs/accelerate/getting-started#23-generate-prisma-client-for-accelerate but on how to generate mappings, not much about what it actually does

further down it mentions "prisma middleware" but if you follow there you get a deprecation warning https://www.prisma.io/docs/orm/prisma-client/client-extensions/middleware but following https://www.prisma.io/docs/orm/prisma-client/client-extensions/query it seems that "prisma middleware" is to inject changes into prisma calls itself and nothing to do with this :)

**Comment by bulltools** [comment]
If you are using next.js 13 you'll have to implement the code as an API route, 

you need to create a new file in the api folder as prisma can't run in the browser

**Reply by sipkoepp** [reply]
What if this api route is protected (e.g. via next-auth)?

**Reply by Babylon-Lion** [reply]
import { PrismaClient } from "@prisma/client"

declare global {
  var prisma: PrismaClient | undefined
}

const client = globalThis.prisma || new PrismaClient()
if (process.env.NODE_ENV !== "production") globalThis.prisma = client

export default client

**Reply by 2mdsajid** [reply]
did it work?

**Reply by m7medVision** [reply]
No

**Comment by elisaado** [comment]
Sadly still not possible, easiest solution seems to be to colocate your middleware logic in rootLayout.

**Reply by frasergr** [reply]
Can you share an example of this?

**Comment by mmahalwy** [comment]
Getting the same error. Any have a fix for this?

**Comment by nemoeric** [comment]
Weird situation. It would be really great to be able to query DB and pass along data inside the request, as it can be done with express.
That help enrich request and/or control access to path. 

Will it work at some point ? Or is it a misconception to structure logic this way ?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
